web/src/message_edit.js

```diff
@@ -967,11 +968,15 @@ export function save_message_row_edit($row) {
                 }
 
                 hide_message_edit_spinner($row);
-                const message = channel.xhr_error_message(
-                    $t({defaultMessage: "Error saving edit"}),
-                    xhr,
+                const message = JSON.parse(xhr.responseText).msg;
+                const $container = compose_banner.get_compose_banner_container(
+                    $row.find("textarea"),
+                );
+                compose_banner.show_error_message(
+                    message,
+                    compose_banner.CLASSNAMES.generic_compose_error,
+                    $container,
                 );
-                $row.find(".edit_error").text(message).show();
```

Comment:
We don't know that the `xhr.responseText` is JSON -- it's an error response, and those are sometimes weird. See `ui_report.error` for some sample code that does the right thing (`xhr_error_message`, the function you're replacing, has similar logic):

```
export function error(                                                                                                                
    response_html: string,                                                                                                            
    xhr: JQuery.jqXHR | undefined,                                                                                                    
    status_box: JQuery,                                                                                                               
    remove_after?: number,                                                                                                            
): void {                                                                                                                             
    if (xhr && xhr.status >= 400 && xhr.status < 500) {                                                                               
        // Only display the error response for 4XX, where we've crafted                                                               
        // a nice response.                                                                                                           
        const server_response_html = _.escape(JSON.parse(xhr.responseText).msg);                                                      
        if (response_html) {                                                                                                          
            response_html += ": " + server_response_html;                                                                             
        } else {                                                                                                                      
            response_html = server_response_html;                                                                                     
        }                                                                                                                             
    }                                                                                                                                 
                                                                                                                                      
    message(response_html, status_box, "alert-error", remove_after);                                                                  
}                                                                                                                                     
                                                                                                                                      
```

Reply:
I have made some tweaks to `xhr_error_message` to allow an optional message parameter, similar to how it's implemented in `ui_report.error`. It seems like we need a unified function to handle the task of returning `response_html`, but I am not sure if I should work on that right now and where to place it.